doc/tutorial/index.rst

```diff
@@ -437,13 +437,74 @@ And make the sentence you added in ``index.rst`` look like this:
    :ref:`install <installation>` the project.
 
 Notice a trick here: the ``install`` part specifies how the link will look like
-(we want it to be a specific word, so the sentence makes sense), whereas the
-``<installation>`` part refers to the actual label we want to add a
+(in this case, using a specific word so the sentence makes sense), whereas the
+``<installation>`` part refers to the actual label you want to add a
 cross-reference to. If you do not include an explicit title, hence using
 ``:ref:`installation```, the section title will be used (in this case,
 ``Installation``). Both the ``:doc:`` and the ``:ref:`` roles will be rendered
 as hyperlinks in the HTML documentation.
 
+Referencing external projects with intersphinx
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You are likely depending on some open source libraries for your own
+project, and there is a chance those dependencies have their documentation
+written in Sphinx and available online. One of the powerful features of
+Sphinx is the ability to generate cross-references to other
+projects available online, using the
+:doc:`intersphinx </usage/extensions/intersphinx>` extension.
+
+To enable it, first add ``sphinx.ext.intersphinx`` to the list of extensions:
+
+.. code-block:: python
+   :caption: docs/source/conf.py
+
+   extensions = [
+       'sphinx.ext.duration',
+       'sphinx.ext.intersphinx',
+   ]
+
+The way to know whether you can reference an external project is to check
+if it has a publicly available ``objects.inv``, which can be parsed calling
```

Comment:
> So I think the way to find the remote objects is also reading source documents, first. Is searching objects.inv really better way to do that? I don't think so...

This was only to explain what the file is. Perhaps it is too technical? I find myself using this quite often lately, perhaps because I'm still learning the intricacies of intersphinx. @webknjaz has https://webknjaz.github.io/intersphinx-untangled/ and I think it's quite useful too.

In any case, I can clarify that this is optional and that it works in the same way as your own docs (except for custom roles, like Sphinx own `:confval:`.

Reply:
> This tutorial does not describe the relationship between the std:doc and :doc: role. Where they came from? I'm worried this explanation makes beginners confused.

It doesn't because I need to learn it myself...!

Reply:
I had no luck on the Sphinx docs but I found this [somewhere else](https://alnoki.readthedocs.io/en/latest/dev-guide/procedures/sphinx.html#id12):

![Screenshot 2021-07-13 at 17-21-41 6 4 Sphinx — AAAAAA 0 4 0 documentation](https://user-images.githubusercontent.com/316517/125479107-dcd3526c-5948-4adf-92ef-dd3561fd69ae.png)

Now I understand your point. I will think if I can clarify this earlier, otherwise rephrase this part.

Reply:
I'd love to see this table in the official docs.

Reply:
I am trying to make sense of domains, and the fact that "Cross-references are generated by many semantic interpreted text roles" ([ref](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#cross-referencing-syntax)) but it's not explicitly stated _which ones_. Then I found that `std:doc` has some special handling:

https://github.com/sphinx-doc/sphinx/blob/5559e5af1ff6f5fc2dc70679bdd6dc089cfff388/sphinx/domains/std.py#L1017-L1018

Reply:
I would strongly prefer to consider ``objects.inv`` an implementation detail that should not go into the tutorial. The current format is quite limiting and it only works for relatively simple reference types, but it needs to be changed in the future to make more complicated cases work (see e.g., https://github.com/sphinx-doc/sphinx/pull/8929). If I remember correctly the ``get_objects()`` method for domains is basically only used for ``objects.inv`` so it will probably get changed as well.

Reply:
> I am trying to make sense of domains, and the fact that "Cross-references are generated by many semantic interpreted text roles" ([ref](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#cross-referencing-syntax)) but it's not explicitly stated _which ones_. Then I found that `std:doc` has some special handling:

It inherently can't state which, as each domain can independently offer such roles, and domains can be third-party extensions. The ``std`` domain is a bit special, bit generally, if domain has a directive to declare an entity, then there usually is a corresponding role for referring to it, e.g., ``.. rst:directive::`` has ``:rst:dir:`` and ``.. py:class::`` has ``:py:class:``.
In this case, if I understand it correctly, the ``std`` considers each document to be an entity implicitly declared, and offers ``:doc:`` to refer them. Similarly, I think when you create a label somewhere, it is really an entity declared in the ``std`` domain, which you can refer to with ``:ref:``. @tk0miya, is this last part correct?

Reply:
Well, after spending some time on this, I think it's better that I don't give `objects.inv` files such a central treatment in this tutorial. Yes, they are useful (I use them all the time now) but the big problem is that we don't have a reference or a cheatsheet where the user can look up all possible keys that will be found there (`std:doc`, `std:label`, the deprecated `std:cmdoption`, and so forth) with the corresponding roles that can be used to insert cross-references to them (`std:doc` ▶️ `:doc:`, `std:label` ▶️ `:ref:`, etc). [Tutorials should only contain minimum, necessary, explanations](https://diataxis.fr/tutorials/#offer-only-minimum-necessary-explanation), but the example I crafted here is just too magical and can lead to frustration. In summary, I will rework this section.

Reply:
It highlights a good point: that we should have better tooling around intersphinx inventories, to help users and hide the implementation details.

Reply:
I have one last caveat:

- In my dummy project, if I enable intersphinx and write

```
:doc:`contents`
```

the link goes to https://www.sphinx-doc.org/en/master/contents.html, as advertised by `objects.inv`.

- However, from Sphinx itself, if I write

```
:doc:`contents`
```

it goes to... https://docs.python.org/3/contents.html! For it to work, I need to write

```
:doc:`/contents`
```

instead. This was very surprising to me, and I think it makes more difficult to introduce the topic: it's not as easy as "write a reference as you would write it in your docs", the user also has to understand some extra rules about how intersphinx targets are generated. Am I missing something @tk0miya @jakobandersen ?

Reply:
(Also, sorry @jakobandersen - I just saw your replies now, somehow the GitHub UI was not showing them to me? Maybe I needed to refresh the page or something 😕)

Reply:
I'm not sure about the leading slash and what is going on there, but you example reminds me of  #8981 which allows to ban the transparent resolution through intersphinx, e.g., as simple typos can suddenly resolve to an external entity instead of an internal one.

Reply:
I'd really like to advocate for explicit remote references for intersphinx. If you view it as namespace, there's lots of ways to have things resolve to something unexpected (as @astrojuanlu mentioned above). I would really like there to be one, well-defined way to reference a document at a specific intersphinx mapping, without any ambiguity as to what I'm referencing. #8981 is a good step towards that 👍

Reply:
> We've usually found local objects via reading source documents, I believe. So I think the way to find the remote objects is also reading source documents, first. Is searching `objects.inv` really better way to do that? I don't think so...
> 

I've never had to do this by reading the source — it's hard to locate such things when you don't know where they are defined and how. Besides, the source is not always linked or available.
This is why I came up with intersphinx-untangled — it helps me discover things that are actually available. I sometimes even search by URL to see what roles link there.

I think there should be some sort of a reference explorer (I mentioned this somewhere on the tracker earlier).

Reply:
> It highlights a good point: that we should have better tooling around intersphinx inventories, to help users and hide the implementation details.

Yep: https://github.com/sphinx-doc/sphinx/issues/8211

Reply:
>In this case, if I understand it correctly, the std considers each document to be an entity implicitly declared, and offers :doc: to refer them. Similarly, I think when you create a label somewhere, it is really an entity declared in the std domain, which you can refer to with :ref:. @tk0miya, is this last part correct?

Yes. Some kinds of objects (including labels and documents) belong to the std domain. It behaves like a system domain. So users do not need to give the domain name on referencing objects like `:std:doc` or `:std:ref:`.
https://github.com/sphinx-doc/sphinx/blob/4.x/sphinx/domains/std.py#L562-L570

Reply:
> This was very surprising to me, and I think it makes more difficult to introduce the topic: it's not as easy as "write a reference as you would write it in your docs", the user also has to understand some extra rules about how intersphinx targets are generated. Am I missing something @tk0miya @ jakobandersen ?

I've never thought referencing remote document via intersphinx. So I did not know such a behavior. Indeed, it would be a bug.

Reply:
> I've never thought referencing remote document via intersphinx. So I did not know such a behavior. Indeed, it would be a bug.

Thanks for chiming in @tk0miya ! I opened a discussion on https://groups.google.com/g/sphinx-dev/c/ZfslSpDGNUs/m/RjfzzFzUCgAJ about it, I think it's somewhat addressed by #9459